src/Lachain.Storage/DbCompact/DbShrinkRepository.cs

```diff
@@ -0,0 +1,118 @@
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using Lachain.Proto;
+using Lachain.Crypto;
+using Lachain.Storage.Trie;
+
+namespace Lachain.Storage.DbCompact
+{
+    public class DbShrinkRepository : IDbShrinkRepository
+    {
+        private IRocksDbContext _dbContext;
+        private RocksDbAtomicWrite batch;
+        private readonly ConcurrentDictionary<UInt256, byte[]> _memDbToSave = new ConcurrentDictionary<UInt256, byte[]>();
```

Comment:
I think maintaining two different hashMaps and underlying assumptions make it less general. Why don't we keep a Map of (hash, List<byte[]>) so that we have all the versions and we can just take the last of the list ?

Reply:
good idea!